File: apps/game-bridge/src/zml_game_bridge/persistence/mining_drops.py
from __future__ import annotations

import sqlite3
from dataclasses import dataclass
from typing import Any, Literal, cast

from zml_game_bridge.domain.mining_events import (
    MiningDropEvent,
    MiningHitHintEvent,
    MiningNoResourcesEvent,
)
from zml_game_bridge.domain.money import Mpec, mpec_to_int
from zml_game_bridge.domain.position import WorldPos
from zml_game_bridge.events.base import EventBase
from zml_game_bridge.events.envelope import EventEnvelope
from zml_game_bridge.persistence.event_projector import EventProjector

MiningDropResult = Literal["pending", "hit", "no_resources"]


@dataclass(frozen=True, slots=True)
class MiningDropRow:
    drop_id: str
    drop_event_id: int
    observed_ts_ms: int
    position: WorldPos | None
    drop_radius_m: float
    modes_mask: int | None
    probes_per_drop: int | None
    ammo_per_drop: int | None
    ammo_cost_mpec: Mpec
    probes_cost_mpec: Mpec
    finder_decay_mpec: Mpec
    finder_enhancer_decay_mpec: Mpec
    amp_decay_mpec: Mpec
    total_cost_mpec: Mpec
    result: MiningDropResult
    result_event_id: int | None
    result_observed_ts_ms: int | None
    hit_id: str | None
    hit_event_id: int | None
    resource_name: str | None
    size_label: str | None
    size_index: int | None
    expected_expires_ts_ms: int | None
    range_m: float | None
    depth_m: float | None


class MiningDropReader:
    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn

    def list_since(self, *, since_ts_ms: int) -> list[MiningDropRow]:
        cur = self._conn.execute(
            """
            SELECT *
            FROM mining_drops
            WHERE observed_ts_ms >= ?
            ORDER BY observed_ts_ms DESC, drop_event_id DESC
            """,
            (since_ts_ms,),
        )
        return [_row_to_mining_drop(row) for row in cur.fetchall()]

    def get(self, drop_id: str) -> MiningDropRow | None:
        cur = self._conn.execute(
            """
            SELECT *
            FROM mining_drops
            WHERE drop_id = ?
            """,
            (drop_id,),
        )
        row = cur.fetchone()
        return _row_to_mining_drop(row) if row is not None else None


class MiningDropProjector(EventProjector):
    def project(
        self,
        *,
        conn: sqlite3.Connection,
        event: EventBase,
        envelope: EventEnvelope,
    ) -> None:
        writer = _MiningDropProjectionWriter(conn)
        if isinstance(event, MiningDropEvent):
            writer.upsert_drop(event=event, event_id=envelope.event_id)
        elif isinstance(event, MiningHitHintEvent):
            writer.mark_hit(event=event, event_id=envelope.event_id)
        elif isinstance(event, MiningNoResourcesEvent):
            writer.mark_no_resources(event=event, event_id=envelope.event_id)


class _MiningDropProjectionWriter:
    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn

    def upsert_drop(self, *, event: MiningDropEvent, event_id: int) -> None:
        position = event.position
        self._conn.execute(
            """
            INSERT INTO mining_drops (
                drop_id, drop_event_id, observed_ts_ms,
                planet_name, x, y, z, drop_radius_m,
                modes_mask, probes_per_drop, ammo_per_drop,
                ammo_cost_mpec, probes_cost_mpec, finder_decay_mpec,
                finder_enhancer_decay_mpec, amp_decay_mpec, total_cost_mpec,
                result
            )
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'pending')
            ON CONFLICT(drop_id) DO UPDATE SET
                drop_event_id = excluded.drop_event_id,
                observed_ts_ms = excluded.observed_ts_ms,
                planet_name = excluded.planet_name,
                x = excluded.x,
                y = excluded.y,
                z = excluded.z,
                drop_radius_m = excluded.drop_radius_m,
                modes_mask = excluded.modes_mask,
                probes_per_drop = excluded.probes_per_drop,
                ammo_per_drop = excluded.ammo_per_drop,
                ammo_cost_mpec = excluded.ammo_cost_mpec,
                probes_cost_mpec = excluded.probes_cost_mpec,
                finder_decay_mpec = excluded.finder_decay_mpec,
                finder_enhancer_decay_mpec = excluded.finder_enhancer_decay_mpec,
                amp_decay_mpec = excluded.amp_decay_mpec,
                total_cost_mpec = excluded.total_cost_mpec
            """,
            (
                event.drop_id,
                event_id,
                event.observed_ts_ms,
                position.planet_name if position is not None else None,
                position.x if position is not None else None,
                position.y if position is not None else None,
                position.z if position is not None else None,
                event.drop_radius_m if event.drop_radius_m is not None else 55.0,
                event.modes_mask,
                event.probes_per_drop,
                event.ammo_per_drop,
                mpec_to_int(event.cost.ammo.cost_mpec),
                mpec_to_int(event.cost.probes.cost_mpec),
                mpec_to_int(event.cost.finder_decay_mpec),
                mpec_to_int(event.cost.finder_enhancer_decay_mpec),
                mpec_to_int(event.cost.amp_decay_mpec),
                mpec_to_int(event.cost.total_mpec),
            ),
        )

    def mark_hit(self, *, event: MiningHitHintEvent, event_id: int) -> None:
        if event.drop_id is None:
            return
        # TODO: This read model stores only the single finder hint for MVP.
        # Real claims/deeds should be projected into a separate claim/deed model,
        # because one multi-mode drop can create more than one deed.
        self._conn.execute(
            """
            UPDATE mining_drops
            SET result = 'hit',
                result_event_id = ?,
                result_observed_ts_ms = ?,
                hit_id = ?,
                hit_event_id = ?,
                resource_name = ?,
                size_label = ?,
                size_index = ?,
                expected_expires_ts_ms = ?,
                range_m = ?,
                depth_m = ?
            WHERE drop_id = ?
            """,
            (
                event_id,
                event.observed_ts_ms,
                event.hit_id,
                event_id,
                event.resource_name,
                event.size_label,
                event.size_index,
                event.expected_expires_ts_ms,
                event.range_m,
                event.depth_m,
                event.drop_id,
            ),
        )

    def mark_no_resources(self, *, event: MiningNoResourcesEvent, event_id: int) -> None:
        if event.drop_id is None:
            return
        self._conn.execute(
            """
            UPDATE mining_drops
            SET result = 'no_resources',
                result_event_id = ?,
                result_observed_ts_ms = ?
            WHERE drop_id = ?
            """,
            (event_id, event.observed_ts_ms, event.drop_id),
        )


def _row_to_mining_drop(row: sqlite3.Row) -> MiningDropRow:
    x = row["x"]
    y = row["y"]
    position = (
        WorldPos(
            planet_name=row["planet_name"],
            x=int(x),
            y=int(y),
            z=int(row["z"]) if row["z"] is not None else None,
        )
        if x is not None and y is not None
        else None
    )
    return MiningDropRow(
        drop_id=str(row["drop_id"]),
        drop_event_id=int(row["drop_event_id"]),
        observed_ts_ms=int(row["observed_ts_ms"]),
        position=position,
        drop_radius_m=float(row["drop_radius_m"]),
        modes_mask=_optional_int(row["modes_mask"]),
        probes_per_drop=_optional_int(row["probes_per_drop"]),
        ammo_per_drop=_optional_int(row["ammo_per_drop"]),
        ammo_cost_mpec=Mpec(int(row["ammo_cost_mpec"])),
        probes_cost_mpec=Mpec(int(row["probes_cost_mpec"])),
        finder_decay_mpec=Mpec(int(row["finder_decay_mpec"])),
        finder_enhancer_decay_mpec=Mpec(int(row["finder_enhancer_decay_mpec"])),
        amp_decay_mpec=Mpec(int(row["amp_decay_mpec"])),
        total_cost_mpec=Mpec(int(row["total_cost_mpec"])),
        result=cast(MiningDropResult, row["result"]),
        result_event_id=_optional_int(row["result_event_id"]),
        result_observed_ts_ms=_optional_int(row["result_observed_ts_ms"]),
        hit_id=row["hit_id"],
        hit_event_id=_optional_int(row["hit_event_id"]),
        resource_name=row["resource_name"],
        size_label=row["size_label"],
        size_index=_optional_int(row["size_index"]),
        expected_expires_ts_ms=_optional_int(row["expected_expires_ts_ms"]),
        range_m=_optional_float(row["range_m"]),
        depth_m=_optional_float(row["depth_m"]),
    )


def _optional_int(value: Any) -> int | None:
    return int(value) if value is not None else None


def _optional_float(value: Any) -> float | None:
    return float(value) if value is not None else None

File: packages/shared/src/dto/miningDrops.ts
import type { WorldPosDTO } from "./worldPos";
import type { MiningClaimCreatedEventWire } from "./miningClaims";

export type MiningDropResult = "pending" | "hit" | "no_resources";

export type MiningDropPositionDto = WorldPosDTO;

export type MiningDropCostDto = {
    ammoCostMpec: number;
    probesCostMpec: number;
    finderDecayMpec: number;
    finderEnhancerDecayMpec: number;
    ampDecayMpec: number;
    totalTtMpec: number;
    totalWithMarkupMpec: number;
};

export type MiningDropDto = {
    dropId: string;
    dropEventId: number;
    runId: number | null;
    segmentId: string | null;
    observedTsMs: number;
    position: MiningDropPositionDto | null;
    dropRadiusM: number;
    modesMask: number | null;
    probesPerDrop: number | null;
    ammoPerDrop: number | null;
    cost: MiningDropCostDto;
    result: MiningDropResult;
    resultEventId: number | null;
    resultObservedTsMs: number | null;
    hitId: string | null;
    hitEventId: number | null;
    resourceName: string | null;
    sizeLabel: string | null;
    sizeIndex: number | null;
    expectedExpiresTsMs: number | null;
    rangeM: number | null;
    depthM: number | null;
};

export type MiningDropPositionWire = {
    planet_name: string | null;
    x: number;
    y: number;
    z?: number | null;
};

export type MiningDropCostWire = {
    ammo_cost_mpec: number;
    probes_cost_mpec: number;
    finder_decay_mpec: number;
    finder_enhancer_decay_mpec: number;
    amp_decay_mpec: number;
    total_tt_mpec: number;
    total_with_markup_mpec: number;
};

export type MiningDropWire = {
    drop_id: string;
    drop_event_id: number;
    run_id?: number | null;
    segment_id?: string | null;
    observed_ts_ms: number;
    position: MiningDropPositionWire | null;
    drop_radius_m: number;
    modes_mask: number | null;
    probes_per_drop: number | null;
    ammo_per_drop: number | null;
    cost: MiningDropCostWire;
    result: MiningDropResult;
    result_event_id: number | null;
    result_observed_ts_ms: number | null;
    hit_id: string | null;
    hit_event_id: number | null;
    resource_name: string | null;
    size_label: string | null;
    size_index: number | null;
    expected_expires_ts_ms: number | null;
    range_m: number | null;
    depth_m: number | null;
};

type MiningDropEventCostComponentWire = {
    quantity: number | null;
    cost_mpec: number;
    source: string;
};

type MiningDropEventCostWire = {
    ammo: MiningDropEventCostComponentWire;
    probes: MiningDropEventCostComponentWire;
    finder_decay_mpec: number;
    finder_enhancer_decay_mpec?: number;
    amp_decay_mpec: number;
    total_tt_mpec: number;
    total_with_markup_mpec: number;
};

export type MiningDropEventWire = {
    drop_id: string;
    run_id?: number | null;
    segment_id?: string | null;
    observed_ts_ms: number;
    position: MiningDropPositionWire | null;
    drop_radius_m: number;
    modes_mask: number | null;
    probes_per_drop: number | null;
    ammo_per_drop: number | null;
    cost: MiningDropEventCostWire;
    raw_status_text?: string | null;
};

export type MiningHitHintEventWire = {
    hit_id: string;
    drop_id: string;
    observed_ts_ms: number;
    position: MiningDropPositionWire | null;
    size_label: string | null;
    size_index: number | null;
    resource_name: string | null;
    expected_expires_ts_ms?: number | null;
    range_m: number | null;
    depth_m: number | null;
};

export type MiningNoResourcesEventWire = {
    drop_id: string;
    observed_ts_ms: number;
    position: MiningDropPositionWire | null;
};

export type AgentEventEnvelope<TType extends string, TPayload> = {
    type: TType;
    eventId?: number;
    createdTsMs: number;
    eventDt: string | null;
    payload: TPayload;
};

export function isMiningDropWire(value: unknown): value is MiningDropWire {
    if (!isRecord(value)) return false;
    return (
        typeof value.drop_id === "string" &&
        isFiniteNumber(value.drop_event_id) &&
        (value.run_id === undefined || isNullableNumber(value.run_id)) &&
        (value.segment_id === undefined || isNullableString(value.segment_id)) &&
        isFiniteNumber(value.observed_ts_ms) &&
        isNullablePositionWire(value.position) &&
        isFiniteNumber(value.drop_radius_m) &&
        isNullableNumber(value.modes_mask) &&
        isNullableNumber(value.probes_per_drop) &&
        isNullableNumber(value.ammo_per_drop) &&
        isMiningDropCostWire(value.cost) &&
        isMiningDropResult(value.result) &&
        isNullableNumber(value.result_event_id) &&
        isNullableNumber(value.result_observed_ts_ms) &&
        isNullableString(value.hit_id) &&
        isNullableNumber(value.hit_event_id) &&
        isNullableString(value.resource_name) &&
        isNullableString(value.size_label) &&
        isNullableNumber(value.size_index) &&
        isNullableNumber(value.expected_expires_ts_ms) &&
        isNullableNumber(value.range_m) &&
        isNullableNumber(value.depth_m)
    );
}

export function isMiningDropEventWire(value: unknown): value is MiningDropEventWire {
    if (!isRecord(value)) return false;
    return (
        typeof value.drop_id === "string" &&
        (value.run_id === undefined || isNullableNumber(value.run_id)) &&
        (value.segment_id === undefined || isNullableString(value.segment_id)) &&
        isFiniteNumber(value.observed_ts_ms) &&
        isNullablePositionWire(value.position) &&
        isFiniteNumber(value.drop_radius_m) &&
        isNullableNumber(value.modes_mask) &&
        isNullableNumber(value.probes_per_drop) &&
        isNullableNumber(value.ammo_per_drop) &&
        isMiningDropEventCostWire(value.cost)
    );
}

export function isMiningHitHintEventWire(value: unknown): value is MiningHitHintEventWire {
    if (!isRecord(value)) return false;
    return (
        typeof value.hit_id === "string" &&
        typeof value.drop_id === "string" &&
        isFiniteNumber(value.observed_ts_ms) &&
        isNullablePositionWire(value.position) &&
        isNullableString(value.size_label) &&
        isNullableNumber(value.size_index) &&
        isNullableString(value.resource_name) &&
        (value.expected_expires_ts_ms === undefined ||
            isNullableNumber(value.expected_expires_ts_ms)) &&
        isNullableNumber(value.range_m) &&
        isNullableNumber(value.depth_m)
    );
}

export function isMiningNoResourcesEventWire(value: unknown): value is MiningNoResourcesEventWire {
    if (!isRecord(value)) return false;
    return (
        typeof value.drop_id === "string" &&
        isFiniteNumber(value.observed_ts_ms) &&
        isNullablePositionWire(value.position)
    );
}

export function wireToMiningDropDto(wire: MiningDropWire): MiningDropDto {
    return {
        dropId: wire.drop_id,
        dropEventId: wire.drop_event_id,
        runId: wire.run_id ?? null,
        segmentId: wire.segment_id ?? null,
        observedTsMs: wire.observed_ts_ms,
        position: wire.position ? wireToPositionDto(wire.position) : null,
        dropRadiusM: wire.drop_radius_m,
        modesMask: wire.modes_mask,
        probesPerDrop: wire.probes_per_drop,
        ammoPerDrop: wire.ammo_per_drop,
        cost: {
            ammoCostMpec: wire.cost.ammo_cost_mpec,
            probesCostMpec: wire.cost.probes_cost_mpec,
            finderDecayMpec: wire.cost.finder_decay_mpec,
            finderEnhancerDecayMpec: wire.cost.finder_enhancer_decay_mpec,
            ampDecayMpec: wire.cost.amp_decay_mpec,
            totalTtMpec: wire.cost.total_tt_mpec,
            totalWithMarkupMpec: wire.cost.total_with_markup_mpec,
        },
        result: wire.result,
        resultEventId: wire.result_event_id,
        resultObservedTsMs: wire.result_observed_ts_ms,
        hitId: wire.hit_id,
        hitEventId: wire.hit_event_id,
        resourceName: wire.resource_name,
        sizeLabel: wire.size_label,
        sizeIndex: wire.size_index,
        expectedExpiresTsMs: wire.expected_expires_ts_ms,
        rangeM: wire.range_m,
        depthM: wire.depth_m,
    };
}

export function miningDropDtoFromMiningDropEventWire(
    wire: MiningDropEventWire,
    eventId: number = -1,
): MiningDropDto {
    return {
        dropId: wire.drop_id,
        dropEventId: eventId,
        runId: wire.run_id ?? null,
        segmentId: wire.segment_id ?? null,
        observedTsMs: wire.observed_ts_ms,
        position: wire.position ? wireToPositionDto(wire.position) : null,
        dropRadiusM: wire.drop_radius_m,
        modesMask: wire.modes_mask,
        probesPerDrop: wire.probes_per_drop,
        ammoPerDrop: wire.ammo_per_drop,
        cost: {
            ammoCostMpec: wire.cost.ammo.cost_mpec,
            probesCostMpec: wire.cost.probes.cost_mpec,
            finderDecayMpec: wire.cost.finder_decay_mpec,
            finderEnhancerDecayMpec: wire.cost.finder_enhancer_decay_mpec ?? 0,
            ampDecayMpec: wire.cost.amp_decay_mpec,
            totalTtMpec: wire.cost.total_tt_mpec,
            totalWithMarkupMpec: wire.cost.total_with_markup_mpec,
        },
        result: "pending",
        resultEventId: null,
        resultObservedTsMs: null,
        hitId: null,
        hitEventId: null,
        resourceName: null,
        sizeLabel: null,
        sizeIndex: null,
        expectedExpiresTsMs: null,
        rangeM: null,
        depthM: null,
    };
}

export function miningDropDtoWithHitHint(
    drop: MiningDropDto,
    wire: MiningHitHintEventWire,
    eventId: number = -1,
): MiningDropDto {
    return {
        ...drop,
        result: "hit",
        resultEventId: eventId,
        resultObservedTsMs: wire.observed_ts_ms,
        hitId: wire.hit_id,
        hitEventId: eventId,
        resourceName: wire.resource_name,
        sizeLabel: wire.size_label,
        sizeIndex: wire.size_index,
        expectedExpiresTsMs: wire.expected_expires_ts_ms ?? null,
        rangeM: wire.range_m,
        depthM: wire.depth_m,
    };
}

export function miningDropDtoWithNoResources(
    drop: MiningDropDto,
    wire: MiningNoResourcesEventWire,
    eventId: number = -1,
): MiningDropDto {
    if (drop.result === "hit") {
        return drop;
    }
    return {
        ...drop,
        result: "no_resources",
        resultEventId: eventId,
        resultObservedTsMs: wire.observed_ts_ms,
    };
}

export function miningDropDtoWithClaimCreated(
    drop: MiningDropDto,
    wire: MiningClaimCreatedEventWire,
    eventId: number = -1,
): MiningDropDto {
    return {
        ...drop,
        result: "hit",
        resultEventId: drop.result === "hit" ? drop.resultEventId : eventId,
        resultObservedTsMs: drop.result === "hit" ? drop.resultObservedTsMs : wire.observed_ts_ms,
        hitId: drop.hitId ?? wire.hit_id,
        resourceName: wire.resource_name ?? drop.resourceName,
        sizeLabel: wire.size_label ?? drop.sizeLabel,
        sizeIndex: wire.size_index ?? drop.sizeIndex,
        expectedExpiresTsMs: wire.expected_expires_ts_ms ?? drop.expectedExpiresTsMs,
        rangeM: wire.range_m ?? drop.rangeM,
        depthM: wire.depth_m ?? drop.depthM,
    };
}

function wireToPositionDto(wire: MiningDropPositionWire): MiningDropPositionDto {
    return {
        planetName: wire.planet_name ?? "",
        x: wire.x,
        y: wire.y,
        z: wire.z ?? null,
    };
}

function isMiningDropCostWire(value: unknown): value is MiningDropCostWire {
    if (!isRecord(value)) return false;
    return (
        isFiniteNumber(value.ammo_cost_mpec) &&
        isFiniteNumber(value.probes_cost_mpec) &&
        isFiniteNumber(value.finder_decay_mpec) &&
        isFiniteNumber(value.finder_enhancer_decay_mpec) &&
        isFiniteNumber(value.amp_decay_mpec) &&
        isFiniteNumber(value.total_tt_mpec) &&
        isFiniteNumber(value.total_with_markup_mpec)
    );
}

function isMiningDropEventCostWire(value: unknown): value is MiningDropEventCostWire {
    if (!isRecord(value)) return false;
    return (
        isCostComponentWire(value.ammo) &&
        isCostComponentWire(value.probes) &&
        isFiniteNumber(value.finder_decay_mpec) &&
        (value.finder_enhancer_decay_mpec === undefined ||
            isFiniteNumber(value.finder_enhancer_decay_mpec)) &&
        isFiniteNumber(value.amp_decay_mpec) &&
        isFiniteNumber(value.total_tt_mpec) &&
        isFiniteNumber(value.total_with_markup_mpec)
    );
}

function isCostComponentWire(value: unknown): value is MiningDropEventCostComponentWire {
    if (!isRecord(value)) return false;
    return isNullableNumber(value.quantity) && isFiniteNumber(value.cost_mpec) && typeof value.source === "string";
}

function isMiningDropResult(value: unknown): value is MiningDropResult {
    return value === "pending" || value === "hit" || value === "no_resources";
}

function isNullablePositionWire(value: unknown): value is MiningDropPositionWire | null {
    return value === null || isPositionWire(value);
}

function isPositionWire(value: unknown): value is MiningDropPositionWire {
    if (!isRecord(value)) return false;
    return (
        isNullableString(value.planet_name) &&
        isFiniteNumber(value.x) &&
        isFiniteNumber(value.y) &&
        (value.z === undefined || isNullableNumber(value.z))
    );
}

function isNullableNumber(value: unknown): value is number | null {
    return value === null || isFiniteNumber(value);
}

function isFiniteNumber(value: unknown): value is number {
    return typeof value === "number" && Number.isFinite(value);
}

function isNullableString(value: unknown): value is string | null {
    return value === null || typeof value === "string";
}

function isRecord(value: unknown): value is Record<string, unknown> {
    return typeof value === "object" && value !== null;
}
